reader.go

```diff
@@ -284,39 +290,65 @@ func (r *reader) run(ctx context.Context, offset int64) {
 		// to the connection we know we'll want to restart from this offset.
 		offset = start
 
+		errcount := 0
 	readLoop:
 		for {
+			if !sleep(ctx, backoff(errcount, 100*time.Millisecond, 10*time.Second)) {
+				conn.Close()
+				return
+			}
+
 			switch offset, err = r.read(ctx, offset, conn); err {
 			case nil:
+				errcount = 0
+				continue
 			case RequestTimedOut:
-			case context.Canceled: // another reader has taken over
+				// Timeout on the kafka side, this can be safely retried.
+			case OffsetOutOfRange:
```

Comment:
what happens if the reader can't keep up, i imagine the reader may lose messages without us noticing.
this could happen if the reader can't consume fast enough and some messages have been GCed by kafka (expired etc..)